### DISCUSSION TITLE
How does brew calculate dependencies?

### DISCUSSION THREAD

**Discussion by mfriedenhagen** [Homebrew/discussions#3704] [Everyday usage] [answered]
Hi,

I looked into the [formula of oath-toolkit](https://github.com/Homebrew/homebrew-core/blob/86ac48865b7d99af728dbf1c7bb9b680daf50ab8/Formula/oath-toolkit.rb#L25) and it only depends on `libxmlsec1`.

Now when I run the following I do see much more dependencies for `oath-toolkit`:
```
❯ diff -u <(brew deps libxmlsec1) <(brew deps oath-toolkit)
--- /dev/fd/11	2022-09-27 12:58:45.000000000 +0200
+++ /dev/fd/12	2022-09-27 12:58:45.000000000 +0200
@@ -1,24 +1,30 @@
 bdw-gc
-ca-certificates
+c-ares
+gdbm
 gettext
 gmp
 gnutls
 guile
-icu4c
+jemalloc
+libev
 libevent
 libffi
 libgcrypt
 libgpg-error
 libidn2
-libnghttp2
 libtasn1
 libtool
 libunistring
 libxml2
-m4
+libxmlsec1
 nettle
+nghttp2
 openssl@1.1
 p11-kit
 pkg-config
+python@3.9
 readline
+sqlite
+tcl-tk
 unbound
+xz
```

Any ideas why that is so?

**Comment by carlocab** [comment]
`brew deps` just traverses the dependency graph until it finds a leaf. The dependency graph is usually specified via `depends_on`, but depending on your OS there might be some implicit dependencies.

What I get from comparing their dependencies is what you expect:
```diff
❯ diff -u <(brew deps libxmlsec1) <(brew deps oath-toolkit)
--- /dev/fd/13 2022-09-28 15:36:53.114639832 +0800
+++ /dev/fd/14       2022-09-28 15:36:53.115426028 +0800
@@ -14,6 +14,7 @@
 libtool
 libunistring
 libxml2
+libxmlsec1
 m4
 nettle
 openssl@1.1
```

**Reply by mfriedenhagen** [reply]
Hi @carlocab, 

thanks for the answer, I am on macOS 12.6. 
I was expecting the same output that you have.

I was wondering as well why `brew deps` did somehow exchange `libnghttp2` by `nghttp2`.  [The formula](https://github.com/Homebrew/homebrew-core/blob/86ac48865b7d99af728dbf1c7bb9b680daf50ab8/Formula/oath-toolkit.rb#L25) only depends on `libxmlsec1`. 

It looks like `brew deps` on my machine does include the `pkg-config` dependencies (which are build-only) as well although I did not specify these.
Maybe executing `diff -u <(brew deps libxmlsec1) <(brew deps --include-build oath-toolkit)` does show "my output" on your machine as well.

When I use `brew deps --tree oath-toolkit`, the output is more reasonable. So probably I have misconfigured something on my machine.

**Reply by gromgit** [reply]
I get the same single-like diff on:
* Intel macOS from Mojave to Monterey
* ARM64 macOS Monterey
* Intel Linux Ubuntu 22.04.1 and Debian 11

so it's almost certainly some local edits to `homebrew/core`, or a third-party tap shadowing `homebrew/core` with similarly-named formulae.

@mfriedenhagen, what's the output of:
```
brew config
brew doctor
brew info libxmlsec1 oath-toolkit
git -C $(brew --repo homebrew/core) status
```

**Comment by carlocab** [comment, marked as answer]
Ah, I see now. When you installed `oath-toolkit`, `libxmlsec1` still depended on `nghttp2` (because there was no `libnghttp2` formula at the time) through `gnutls` and `unbound`. Hence, `brew` remembers that `oath-toolkit` had a dependency on `nghttp2` (and all its recursive dependencies: `jemalloc`, `libev`, etc) at the point of installation.

You can force `brew` to "forget" that dependency by doing `brew reinstall oath-toolkit`.

**Reply by mfriedenhagen** [reply]
So basically brew does not calculate this kind of information on the fly but during installation, right? Do you know where thins state is kept?

**Reply by iMichka** [reply]
In /usr/local/Cellar/oath-toolkit/--version--/INSTALL_RECEIPT.json

This file is either provided by the bottle, or recreated when building from source.

The content of this file influence `brew deps`.

**Reply by mfriedenhagen** [reply]
Hi @iMichka, 

thanks, that is where the output from `brew deps` stems from:
```
❯ jq -r '.runtime_dependencies | .[] | .full_name' < /usr/local/opt/oath-toolkit/INSTALL_RECEIPT.json | sort
bdw-gc
c-ares
gdbm
gettext
gmp
gnutls
guile
jemalloc
libev
libevent
libffi
libgcrypt
libgpg-error
libidn2
libtasn1
libtool
libunistring
libxml2
libxmlsec1
nettle
nghttp2
openssl@1.1
p11-kit
pkg-config
python@3.9
readline
sqlite
tcl-tk
unbound
xz
```

So: the output from `brew deps` and `brew deps --tree` do use different mechanisms, right? For the tree view only libxmlsec1 is reported:
```
❯ brew deps --tree oath-toolkit
oath-toolkit
└── libxmlsec1
    ├── gnutls
    │   ├── ca-certificates
    │   ├── gmp
    │   ├── guile
    │   │   ├── bdw-gc
    │   │   ├── gmp
    │   │   ├── libtool
    │   │   │   └── m4
    │   │   ├── libunistring
    │   │   ├── pkg-config
    │   │   └── readline
    │   ├── libidn2
    │   │   ├── gettext
    │   │   └── libunistring
    │   ├── libtasn1
    │   ├── libunistring
    │   ├── nettle
    │   │   └── gmp
    │   ├── p11-kit
    │   │   ├── ca-certificates
    │   │   └── libtasn1
    │   └── unbound
    │       ├── libevent
    │       │   └── openssl@1.1
    │       │       └── ca-certificates
    │       ├── libnghttp2
    │       └── openssl@1.1
    │           └── ca-certificates
    ├── libgcrypt
    │   └── libgpg-error
    ├── libxml2
    │   ├── icu4c
    │   └── readline
    └── openssl@1.1
```

**Reply by mfriedenhagen** [reply]
In a ideal world I would mark this as **answer**_

> In /usr/local/Cellar/oath-toolkit/--version--/INSTALL_RECEIPT.json
> 
> This file is either provided by the bottle, or recreated when building from source.
> 
> The content of this file influence `brew deps`.

and 

> Ah, I see now. When you installed `oath-toolkit`, `libxmlsec1` still depended on `nghttp2` (because there was no `libnghttp2` formula at the time) through `gnutls` and `unbound`. Hence, `brew` remembers that `oath-toolkit` had a dependency on `nghttp2` (and all its recursive dependencies: `jemalloc`, `libev`, etc) at the point of installation.
> 
> You can force `brew` to "forget" that dependency by doing `brew reinstall oath-toolkit`.

as **solution** for getting rid of the outdated information. Thanks to all which took part in the discussion :-)
